src/markdown/generator.ml

```diff
@@ -0,0 +1,327 @@
+    | String s -> Format.fprintf fmt "%s" s
+    | Env (indent, s, t, s') ->
+        Format.fprintf fmt "@[%s%s%a@]@,%s@]"
+          (match indent with Any -> "")
+          s pp t s'
+    | Command (s, t) -> Format.fprintf fmt "@[%s%a@]" s pp t
+    | OpenSqBracket -> Format.fprintf fmt "["
+    | CloseSqBracket -> Format.fprintf fmt "]"
+    | OpenParenthesis -> Format.fprintf fmt "("
+    | CloseParenthesis -> Format.fprintf fmt ")"
+
+  and pp_many fmt l = List.iter (pp fmt) l
+end
+
+open Markup
+
+let entity e =
+  match e with "#45" -> escaped "-" | "gt" -> str ">" | s -> str "&%s;" s
+
+let raw_markup (_ : Raw_markup.t) = noop
+
+let style (style : style) content =
+  match style with
+  | `Bold -> command "**" (content ++ str "**")
+  | `Italic | `Emphasis -> command "_" (content ++ str "_")
+  | `Superscript -> command "<sup>" content
+  | `Subscript -> command "<sub>" content
+
+let generate_links = ref false
+
+let rec source_code (s : Source.t) =
+  match s with
+  | [] -> noop
+  | h :: t -> (
+      match h with
+      | Source.Elt i -> inline i ++ source_code t
+      | Tag (None, s) -> source_code s ++ source_code t
+      | Tag (Some _, s) -> source_code s ++ source_code t)
+
+and inline (l : Inline.t) =
+  match l with
+  | [] -> noop
+  | i :: rest -> (
+      match i.desc with
+      | Text "" -> inline rest
+      | Text _ ->
+          let l, _, rest =
+            Doctree.Take.until l ~classify:(function
+              | { Inline.desc = Text s; _ } -> Accum [ s ]
+              | _ -> Stop_and_keep)
+          in
+          str {|%s|} (String.concat "" l) ++ inline rest
+      | Entity e ->
+          let x = entity e in
+          x ++ inline rest
+      | Styled (sty, content) -> style sty (inline content) ++ inline rest
+      | Linebreak -> break ++ inline rest
+      | Link (href, content) ->
+          if !generate_links then
+            (let rec f (content : Inline.t) =
+               match content with
+               | [] -> noop
+               | i :: rest -> (
+                   match i.desc with
+                   | Text s ->
+                       command "" (str "[%s]" s ++ str "(%s)" href) ++ f rest
+                   | _ -> noop ++ f rest)
+             in
+             f content)
+            ++ inline rest
+          else inline content ++ inline rest
+      | InternalLink (Resolved (link, content)) ->
+          if !generate_links then
+            match link.page.parent with
+            | Some _ -> inline content ++ inline rest
+            | None ->
+                open_sq_bracket ++ inline content ++ close_sq_bracket
+                ++ open_parenthesis
+                ++ String ("#" ^ link.anchor)
+                ++ close_parenthesis ++ inline rest
+          else inline content ++ inline rest
+      | InternalLink (Unresolved content) -> inline content ++ inline rest
+      | Source content ->
+          env Any "`` " "`` " (source_code content) ++ inline rest
```

Comment:
Should be `"`` " " ``"` (eg. spaces towards the inside) but most importantly, it should be just `` "`" "`" `` if `content` doesn't contain any `` ` ``. Also, shouldn't it be `command` instead of `env` ? (not sure what `command` does)

Perhaps there should be a `Inline_source` constructor in `Markup` for handling this ? (eg. let that module handle the `` ` ``).

Reply:
instance of https://github.com/ocaml/odoc/pull/723#discussion_r712841985